### ISSUE TITLE
Do not base interface language on LC_TIME

### ISSUE THREAD

**Issue by aostruszka** [slgobinath/SafeEyes#230]
https://github.com/slgobinath/SafeEyes/blob/ee078e5f3bbce4df4927dc4b4d813131d1f2b6b4/safeeyes/Utility.py#L100

I've got maybe non-standard locale settings since I'm from Poland but I prefer to have interface in English.  So I have LANG/LANGUAGE set to en_US.utf8/en_US but I set "LC_CTYPE=pl_PL.utf8 LC_COLLATE=pl_PL.utf8" in my dot files.  In default settings generated by the system (/etc/default/locale) I see also LC_TIME set to pl_PL.utf8 so I can change that but actually I prefer having 24h clock instead of AM/PM.

So I'd like to ask for possibility to:
  1. Base the interface language used on LANG/LC_MESSAGES instead of LC_TIME
  1. Advanced option would be to split locale: having one for time and one for interface/messages (maybe this option is not even needed since correctly installed locale should be respecting LC_TIME override)

Best regards
Andrzej

PS. If you agree that such modification make sense (to me it does - since MESSAGES seems to be more specific to tell what kind of language I want to see in the interface) then it looks to me like the following patch achieves both points:
```
--- Utility-orig.py	2018-01-15 11:56:09.990407876 +0100
+++ Utility.py	2018-01-15 12:02:31.688636092 +0100
@@ -91,13 +91,13 @@
         GLib.idle_add(target_function)
 
 
-def system_locale():
+def system_locale(category=locale.LC_MESSAGES):
     """
     Return the system locale. If not available, return en_US.UTF-8.
     """
     try:
         locale.setlocale(locale.LC_ALL, '')
-        sys_locale = locale.getlocale(locale.LC_TIME)[0]
+        sys_locale = locale.getlocale(category)[0]
         if not sys_locale:
             sys_locale = 'en_US.UTF-8'
         return sys_locale
@@ -110,7 +110,7 @@
     """
     Format time based on the system time.
     """
-    sys_locale = system_locale()
+    sys_locale = system_locale(locale.LC_TIME)
     return babel.dates.format_time(time, format='short', locale=sys_locale)
```

**Comment by slgobinath** [comment]
Hi,
The same issue is reported in https://github.com/slgobinath/SafeEyes/issues/225. Did you check your suggested patch? If it is working as expected, could you please send a PR?

Thanks

**Comment by aostruszka** [comment]
It is working for me (I've asked others in #225 to test) - if it is working for others then I can create PR, but I do not know your app well so there might be other points to be fixed.  For me I get what I want - interface in English and time in 24h format.

**Comment by slgobinath** [comment]
After their confirmation, please send a PR.
If there are any other issues, I will take care of them.

Thanks

**Comment by leedoyle** [comment]
@aostruszka if you don't mind me asking, please send the instructions on how to apply the patch.

**Comment by hartwork** [comment]
Hi!

The patch looks like the right direction. However I believe there is more wrong here: .getlocale returns a 2-tuple of strings:
```
In [5]: locale.getlocale(locale.LC_TIME)
Out[5]: ('de_DE', 'UTF-8')
```
So function `system_locale` can return string `'en_US.UTF-8'` or 2-tuples like `('en_US', 'UTF-8')`. That looks like an unintended mix to me.

**Comment by leedoyle** [comment]
I solved the problem by specifying both the locale and time locale. I want safeeyes to use a locale that is not my system locale, so the 	aostruszka's patch doesn't meet my requirements.

**Comment by leedoyle** [comment]
I find that this should be configurable **at the very least** by editing a .config file, not the program's source code.

**Comment by slgobinath** [comment]
Okay, I will follow from this and fix the issue with a configurable alternative.
I am little busy this week. I will fix this issue in the coming weekend.

**Comment by hartwork** [comment]
> I want safeeyes to use a locale that is not my system locale, so the aostruszka's patch doesn't meet my requirements.

@leedoyle for a workaround you could start safeeyes like `LC_TIME=en_US.utf8 safeeyes` on the shell or adjust a copy of `safeeyes.desktop` put to `~/.local/share/applications/`, I believe. Tested former so far.

**Comment by leedoyle** [comment]
@hartwork well, here's how I did it (quick and dirty)
def system_locale():
    """
    Return the system locale. If not available, return en_US.UTF-8.
    """
    try:
        locale.setlocale(locale.LC_ALL, '')
#        sys_locale = locale.getlocale(locale.LC_TIME)[0]
        if not sys_locale:
            sys_locale = 'de_DE.UTF-8'
        return sys_locale
    except BaseException:
        # Some systems does not return proper locale
        return 'de_DE.UTF-8'


def format_time(time):
    """
    Format time based on the system time.
    """
    sys_locale = 'ru_RU.UTF8'
#    sys_locale = system_locale(locale.LC_TIME)
    return babel.dates.format_time(time, format='short', locale=sys_locale)

**Comment by leedoyle** [comment]
@hartwork btw, it would be nice if you could take a look at /usr/lib/python3/dist-packages/safeeyes/config/locale/de/LC_MESSAGES/safeeyes.po , I have noticed one typo (verzögernn statt verzögern) so far, but maybe there's more.

**Comment by aostruszka** [comment]
@leedoyle 
You are hardcoding the result you want.  Try instead revert the `Utility.py` to original state, copy and paste the patch I've included into some file (say `safe_eyes.patch`) then in the directory with `Utility.py` execute: `patch < safe_eyes.patch`.  This will allow to configure settings via environment variables.

You can test what results you'll get by playing with python shell.  Just start it with different values for LANG/LC_TIME (e.g. `LANG=ru_RU.utf8 LC_TIME=de_DE.utf8 python` for Russian interface with German time formatting) and issue:
```
import locale
locale.setlocale(locale.LC_ALL, '')
```
and observe the values of LC_MESSAGES and LC_TIME.

For example I have LANG=en_US.utf8 (which is the system setting).  This governs all categories unless overridden by more specific settings - and in my case I set LC_CTYPE, LC_COLLATE and LC_TIME to pl_PL.utf8.  In that case LC_MESSAGES gets value from LANG (it could be overridden with explicit LC_MESSAGES= assignment) and LC_TIME from explicit setting and I have interface in English but time in 24h (instead of AM/PM).

So if you have system locale X then you should have LANG=X.  If you want safeeyes to be in:
  - Y with time in Z -> LC_MESSAGES=Y LC_TIME=Z
  - Y with time in X -> LC_MESSAGES=Y
  - X with time in Y -> LC_TIME=Y
  - and so on...

So for me this works as I expect.  LANG/LC_* settings configure locale for my whole session and safeeyes (with the patch) adapt to it.  However if you want safeeyes to have different settings from the rest of your session then you would need to ask @slgobinath to implement it :).

Best regards
Andrzej

**Comment by slgobinath** [comment]
Hi all,
Sorry for my delay due to my busy schedule.
According to my testings, @aostruszka's patch resolves the issue.
If you run Safe Eyes (master branch) using the following command, Safe Eyes should use Russian for interface and German for time format.
```
LANG=ru_RU.utf8 LC_TIME=de_DE.utf8 python3 -m safeeyes
```
I find it more generic and the standard way to go.
Do we still need a configuration to adjust them? WDYT?

**Comment by hartwork** [comment]
> Do we still need a configuration to adjust them? WDYT?

With the fix in, I won't need another way to configure locales, personally.

**Comment by slgobinath** [comment]
Thanks for all your support. I have released Safe Eyes 2.0.3 including this fix.

**Comment by slgobinath** [comment]
@hartwork Please note that I have immediately released 2.0.4 to fix another bug I have noticed after releasing 2.0.3.
